proposals/0000-borrow-operator.md

```diff
@@ -0,0 +1,189 @@
+
+-->
+
+## Introduction
+
+In this document, we propose adding a new operator to the language, marked by
+the context sensitive keyword `borrow`, that causes the compiler to pass
+arguments using a `pass-by-borrow` convention instead of a `pass-by-value`
+convention.
+
+Swift-evolution thread: [Discussion thread topic for that proposal](https://forums.swift.org/)
+
+## Motivation
+
+Swift normally passes arguments using a `pass-by-value` convention. A
+`pass-by-value` argument convention implies that each argument is passed as a
+separate value to the callee. This has the effect that the compiler is allowed
+(and sometimes required) to introduce temporary copies of arguments into the
+program in caller callsites.
+
+As an elucidating example, consider the following code:
+
+```
+func useValueTwice(_ x: Type, _ y: Type) {}
+
+var a: Type = ...
+useValueTwice(a, a)
+```
+
+Since `useValueTwice` is using default Swift conventions, we pass `a` as two
+separate guaranteed parameters to `useValueTwice`. Inside `useValueTwice`, `x`
+and `y` both appears to be separate truly borrowed values in the sense that
+`useValueTwice` has immutable access to two different values and does not take
+ownership of those values. In order for the compiler to achieve this, it must
+necessarily copy `a` to ensure that it can pass two different values to
+`useValueTwice`. These copies are observable program behavior and are in fact a
+desired default behavior since otherwise in general cases we couldn't write code
+like the above that takes the same argument multiple times. We instead would
+necessarily require the compiler to insert an error similar to an exclusivity
+violation which is undesireable:
+
+```
+useValueTwiceWithoutPassByValue(a, a) // Error! Cannot pass a value twice!
+```
+
+In certain cases though, this `pass-by-value` behavior and the implicit copies
+it requires the compiler to insert may be undesireable due to the author needing
+strict performance guarantees that implicit copies like the above are never
+generated. For most programs this guarantee is not needed since the optimizer
+generally does a good enough job eliminating such copies. But in critical system
+code, good enough is not an acceptable constraint and a true semantic guarantee
+is needed.
+
+## Proposed solution
+
+We propose introducing a new context sensitive keyword called `borrow` that when
+applied to a function argument forces Swift to use a `pass-by-borrow` convention
+on a specific argument instead of a `pass-by-value` convention. This would thus
+forbid the compiler from inserting such copies and instead cause the compiler to
+emit an error in such a case:
+
+```
+var a: Type = ...
+useValueTwice(borrow a, borrow a) // Error! Cannot overlapping borrows of 'a'!
+```
+
+The result of the usage of borrow is that a performance minded programmer would
+be able to sleep well at night knowing that their code doesn't have copies in
+this code.
+
+## Detailed design
```

Comment:
Overall, there's a lot of swift-compiler-specific jargon that needs to either be explained or revised to use terms that an average compiler person who has not worked on the Swift compiler can understand. like:

- "guaranteed parameter"
- "a borrowed formal access"

>one will also achieve an exclusivity violation since one will attempt to access the `var`/`inout` for the lvalue-to-rvalue conversion /after/ the formal evaluation of the borrowed argument has begun. Example:

none of these words are in the bible :) .... more seriously, consider whether details like this are actually important for the community to understand when evaluating this _language_ proposal. Esoteric details that require an understanding of concepts specific to the _implementation_ are irrelevant unless if your proposal requires a major overhaul of the compiler.